apis/goharbor.io/v1alpha2/trivy_types.go

```diff
@@ -0,0 +1,263 @@
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Pattern="https?://.+"
+	// The URL of the HTTPS proxy server
+	HTTPSProxy string `json:"httpsProxy,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// The URLs that the proxy settings do not apply to
+	NoProxy []string `json:"noProxy,omitempty"`
+}
+
+// +kubebuilder:validation:Enum={"os","library"}
+// +kubebuilder:default={"os","library"}
+// TrivyServerVulnerabilityType represents a CVE vulnerability type for trivy.
+type TrivyServerVulnerabilityType string
+
+// +kubebuilder:validation:Enum={"UNKNOWN","LOW","MEDIUM","HIGH","CRITICAL"}
+// +kubebuilder:default={"UNKNOWN","LOW","MEDIUM","HIGH","CRITICAL"}
+// TrivyServerSeverityType represents a CVE severity type for trivy.
+type TrivyServerSeverityType string
+
+var trivyURLValidationRegexp = regexp.MustCompile(`https?://.+`)
+
+func (r *TrivyServerSpec) Validate() map[string]error {
+	errors := map[string]error{}
+
+	if len(r.NoProxy) > 0 {
+		for _, url := range r.NoProxy {
+			matched := trivyURLValidationRegexp.MatchString(url)
+			if !matched {
+				errors["NoProxy"] = ErrWrongURLFormat
+				break
+			}
+		}
+	}
+
+	return errors
+}
+
+type TrivyLogSpec struct {
+	// +kubebuilder:validation:Optional
+	Level harbormetav1.TrivyLogLevel `json:"level,omitempty"`
+}
+
+type TrivyHTTPSSpec struct {
+	// +kubebuilder:validation:Optionnal
+	// Reference to secret containing tls certificate
+	CertificateRef string `json:"certificateRef"`
+
+	// +kubebuilder:validation:Optionnal
+	// Reference to secret containing tls key
+	KeyRef string `json:"keyRef"`
+
+	// +kubebuilder:validation:Optionnal
+	// A list of absolute paths to x509 root certificate authorities
+	// that the api use if required to verify a client certificate
+	ClientCas string `json:"clientCasList"`
+}
+
+type TrivyCacheSpec struct {
+	// +kubebuilder:validation:Required
+	// Redis cache store
+	Redis OpacifiedDSN `json:"redis,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="harbor.scanner.trivy:store"
+	// The namespace for keys in the Redis store
+	RedisNamespace string `json:"redisNamespace,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Type="string"
+	// +kubebuilder:default="1h"
+	// +kubebuilder:validation:Pattern="([0-9]+h)?([0-9]+m)?([0-9]+s)?([0-9]+ms)?([0-9]+us)?([0-9]+µs)?([0-9]+ns)?"
+	// The time to live for persisting scan jobs and associated scan reports
+	RedisScanJobTTL *metav1.Duration `json:"redisScanJobTTL,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="harbor.scanner.trivy:job-queue"
+	// The namespace for keys in the scan jobs queue backed by Redis
+	QueueRedisNamespace string `json:"queueRedisNamespace,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1
+	// The number of workers to spin-up for the scan jobs queue
+	QueueWorkerConcurrency int `json:"queueWorkerConcurrency,omitempty"`
```

Comment:
We expect that the `trivy.db` file is accessed by a single `trivy`<sup>[1](https://github.com/aquasecurity/trivy)</sup> process. Increasing the number of concurrent workers may corrupt `trivy.db` and lead to inconsistent state. I wish I have never added this configuration param. TBH I'm thinking about removing it from the adapter service. Then the only way to scale up the scanner adapter would be to scale the Deployment created by the operator.

Reply:
I am removing this option from the PR